mod/state-transition/pkg/core/deposits_validation.go

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2024, Berachain Foundation. All rights reserved.
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package core
+
+import (
+	"fmt"
+
+	"github.com/berachain/beacon-kit/mod/config/pkg/spec"
+	"github.com/berachain/beacon-kit/mod/primitives/pkg/math"
+)
+
+func (sp *StateProcessor[
+	_, _, _, BeaconStateT, _, DepositT,
+	_, _, _, _, _, _, _, _, _, _, _,
+]) validateGenesisDeposits(
+	st BeaconStateT,
+	deposits []DepositT,
+) error {
+	switch {
+	case sp.cs.DepositEth1ChainID() == spec.BartioChainID:
+		// Bartio does not properly validate deposits index
+		// We skip checks for backward compatibility
+		return nil
+
+	case sp.cs.DepositEth1ChainID() == spec.BoonetEth1ChainID:
+		// Boonet inherited the bug from Bartio and it may have added some
+		// validators before we activate the fork. So we skip validation
+		// before fork activation
+		return nil
+
+	default:
+		if _, err := st.GetEth1DepositIndex(); err == nil {
```

Comment:
what error is exactly returned here? Should we explicity check that the error contains "not found" ?

Reply:
Can't do now without importing cosmos sdk module here, which we really don't want. I wrote a comment about this but maybe got lost in updates. Let me find it

Reply:
For the very first deposit, where `Eth1DepositIndex` has not been set yet, we have a ErrNotFound kind of error. This is fine (we haven't even set a single deposit) but it should only happen for the very first deposit (so the check on deposit index)
On the other hand, I agree that we should check the kind of error we receive, and we should behave like that only for ErrNotFound and not others but:
- ErrNotFound is a cosmosSDK error and we don't want that dependency here (discussed with Dev at some point)
- We do this in another place (see `applyDeposit` and how we do create validators
- I have an old PR that start fixing it (https://github.com/berachain/beacon-kit/pull/2065) and I am willing to refresh that and resubmit for review as soon as this in in